lib/oa/graph.rb

```diff
@@ -78,30 +78,17 @@ def id_as_url
     # @return [Array<String>] Array of urls expressing the OA motivated_by
     #   values
     def motivated_by
-      motivations = []
       q = self.class.anno_query.dup
       q << [:s, RDF::Vocab::OA.motivatedBy, :motivated_by]
-      solution = @graph.query q
-      if solution && solution.size > 0
-        solution.each { |res|
-          motivations << res.motivated_by.to_s
-        }
-      end
-      # TODO:  raise exception if none? (validation)
-      motivations
+      # if the motivation object gives junk with .to_s, then the anno is illegal anyway and we're not responsible
+      @graph.query(q).each.collect {|stmt| stmt.motivated_by.to_s}
     end
 
     # @param [RDF::URI] predicate either RDF::Vocab::OA.hasTarget or
     #   RDF::Vocab::OA.hasBody
     # @return [Array<String>] urls for the predicate, as an Array of Strings
     def predicate_urls(predicate)
-      urls = []
-      predicate_solns = @graph.query [nil, predicate, nil]
-      predicate_solns.each { |predicate_stmt|
-        predicate_obj = predicate_stmt.object
-        urls << predicate_obj.to_str.strip if predicate_obj.is_a?(RDF::URI)
-      }
-      urls
+      @graph.query([nil, predicate, nil]).each_object.collect {|o| o.to_str.strip if o.is_a?(RDF::URI) }.compact
```

Comment:
Again, using collect is simpler, that's great, although let's have an issue for the `o.to_str` etc.